cassiopeia/solver/DistanceSolver.py

```diff
@@ -24,122 +25,74 @@ class DistanceSolverError(Exception):
-        meta_data: Data table storing meta data for each sample
-        priors: Prior probabilities of character state transitions
-        weights: Weights on character/mutation pairs, derived from priors
-        dissimilarity_map: Dissimilarity map describing distances between
-            samples
-        dissimilarity_function: Function to compute the dissimilarity between
-            samples.
-        tree: The tree built by `self.solve()`. None if `solve` has not been
-            called yet
-        unique_character_matrix: A character matrix with duplicate rows filtered out
-    """
-
-    def __init__(
-        self,
-        character_matrix: pd.DataFrame,
-        missing_char: int = -1,
-        meta_data: Optional[pd.DataFrame] = None,
-        priors: Optional[Dict] = None,
-        prior_transformation: Optional[
-            Callable[[float], float]
-        ] = "negative_log",
-        dissimilarity_map: Optional[pd.DataFrame] = None,
-        dissimilarity_function: Optional[Callable] = None,
-    ):
-
-        if dissimilarity_function is None and dissimilarity_map is None:
-            raise DistanceSolverError(
-                "Please specify a dissimilarity map or dissimilarity function"
-            )
+    
+    def __init__(self, dissimilarity_function: Optional[Callable] = None):
 
-        super().__init__(character_matrix, meta_data, priors)
+        super().__init__()
 
-        self.tree = None
-        self.weights = None
-
-        if priors:
-            self.weights = solver_utilities.transform_priors(
-                priors, prior_transformation
-            )
-
-        self.dissimilarity_map = dissimilarity_map
         self.dissimilarity_function = dissimilarity_function
 
-        self.unique_character_matrix = (
-            self.character_matrix.drop_duplicates().copy()
-        )
-        self.unique_character_matrix.index = [
-            f"state{i}" for i in range(self.unique_character_matrix.shape[0])
-        ]
-
-        # Create the dissimilarity map if not specified
-        if self.dissimilarity_map is None:
-            N = self.unique_character_matrix.shape[0]
-            dissimilarity_map = self.compute_dissimilarity_map(
-                self.unique_character_matrix.to_numpy(), N
-            )
-            dissimilarity_map = scipy.spatial.distance.squareform(
-                dissimilarity_map
-            )
-
-            self.dissimilarity_map = pd.DataFrame(
-                dissimilarity_map,
-                index=self.unique_character_matrix.index,
-                columns=self.unique_character_matrix.index,
-            )
-
-    def solve(self):
+    def solve(
+        self,
+        cassiopeia_tree: CassiopeiaTree,
+        dissimilarity_map: Optional[pd.DataFrame] = None,
+        root_sample: Optional[str] = None,
+        root_tree: bool = False,
+        prior_transformation: str = "negative_log",
+    ) -> None:
```

Comment:
What are the pros and cons of having the arguments `dissimilarity_map , root_sample, root_tree , prior_transformation` be part of the `solve` method instead of specifying them some other way, e.g. through `__init__` or the `CassiopeiaTree`? One con of having them be part of `solve` is that it violates the uniform `solve` interface we want. As it stands this is not a type error because the arguments are all optional, but this looks weird to me. I realize that `root_sample` is a `cassiopeia_tree`-specific notion, so it would be ridiculous to pass `root_sample` to `__init__` and then `cassiopeia_tree` to `solve`. I agree with this. I don't know yet what the best solution is. But to this end, let me ask: how are the `dissimilarity_map`, `root_sample`, `root_tree`  and `prior_transformation` created?

BTW this is not a red flag to me, I am happy to merge with this as-is but want to understand from a design perspective if we are going in the right direction.

Reply:
I think we're totally on the same page regarding this. It feels a bit funky to me as well, truth be told. The main argument I have for including these as arguments to the `solve` function is that I don't think they are general enough to the algorithm to be included in the __init__ statement.

A simple example is if you have a panel of character matrices you'd like to all reconstruct with the same `DistanceSolver` routine. They'd all have specific dissimilarity maps, priors, character matrices, etc. So it wouldn't make sense to pass things like a `dissimilairty_map` or sample to treat as a root (i.e., `root_sample`) to the __init__ statement because it wouldn't apply generally.

Of course, I anticipate you'd agree here - this was the motivation for feeding in a `CassiopeiaTree` object to the `solve` method. It can be argued that we should extend the `CassiopeiaTree` class to contain a `dissimilarity_map` and putative root sample for input into a `DistanceSolver`. I'm totally happy to do this if you think that would be cleaner. On this note, I also suppose we should have `root_tree` be an argument to the class itself -- this actually _is_ a parameter of the `DistanceSolver`, right? I guess a use-case to consider is if you are comparing the effects of different rooting strategies. These would constitute different solvers, I think (take for example UPGMA which by definition returns a rooted tree).

Regarding the `prior_transformation`, I'm actually quite unsure where's the best place to put this. It could arguably go into the CassiopeiaTree itself, but it might be out of sight for a user when turning to a solver. I prefer having it as a parameter to `solve` because it's a data-specific operation that you might want to specify for a particular tree, in a particular routine, but perhaps not be married to the tree.

Reply:
Yes I think we are on the same page. Regarding data-specific objects that don't quite fit into the solver's `__init__`, and also not into the `CassiopeiaTree`, I'd only like to point out that the solver's `__init__` can alternatively take in a _function_ that computes the data-specific object. As a concrete example say that the `DistanceSolver` needs the root of the tree to be specified. Instead of:

```
class DistanceSolver(CassiopeiaSolver.CassiopeiaSolver):
    def __init__(
        self,
    ):
        ...

    def solve(tree: CassiopeiaTree, root: str):
        ...
```

and **also instead of**:

```
class DistanceSolver(CassiopeiaSolver.CassiopeiaSolver):
    def __init__(
        self,
        root: str
    ):
        ...

    def solve(tree: CassiopeiaTree):
        ...
```

one can have:
```
class DistanceSolver(CassiopeiaSolver.CassiopeiaSolver):
    def __init__(
        self,
        root_finding_algorithm: Callable[[CassiopeiaTree], str]
    ):
        ...

    def solve(tree: CassiopeiaTree):
        root = self.root_finding_algorithm(tree)
        ...
```

Anyway, the key is that this is the spirit of building objects by composition: The `DistanceSolver` becomes a composition of algorithms, one for specifying the root of the tree, one for computing prior transformations, etc. etc.

Reply:
This is a really cool suggestion! I'm not exactly sure how we would implement this, but I'll leave this as a TODO for the DistanceSolver class.